src/number_theory/number_field/basic.lean

```diff
@@ -127,11 +130,81 @@ begin
     (((is_integral_closure.is_integral_algebra ℤ K).is_field_iff_is_field h_inj).mpr hf)
 end
 
-instance [number_field K] : is_dedekind_domain (𝓞 K) :=
+instance : is_dedekind_domain (𝓞 K) :=
 is_integral_closure.is_dedekind_domain ℤ ℚ K _
 
+variables (R : Type*) [comm_ring R] [algebra R K]
+
+lemma _root_.number_field.is_integral_closure.free (hc : is_integral_closure R ℤ K) :
+  free ℤ R :=
+begin
+  haveI : no_zero_smul_divisors ℤ R := function.injective.no_zero_smul_divisors _
+    hc.algebra_map_injective (map_zero _) (map_zsmul _),
+  haveI : is_noetherian ℤ R := is_integral_closure.is_noetherian _ ℚ K _,
+  have basis : Σ n, basis (fin n) ℤ R := free_of_finite_type_torsion_free',
+  obtain ⟨n, b⟩ := basis,
+  exact free.of_basis b,
```

Comment:
`obtain` does strictly more than `have`, so it suffices to write:
```suggestion
  obtain ⟨n, b⟩ : Σ n, basis (fin n) ℤ R := free_of_finite_type_torsion_free',
  exact free.of_basis b,
```
I'm actually surprised that `module.free_of_finite_type_torsion_free` doesn't actually return something of the form `module.free`. So I'd suggest changing that definition to have type `module.free R M` instead of `Σ n, basis (fin n) R M`.

Reply:
Well, I think these results are quite useful and we should keep them. What about I rename `module.free_of_finite_type_torsion_free` (and the `'` version) to `module.basis_of_finite_type_torsion_free` and the new version of `module.free_of_finite_type_torsion_free` states `module.free R M` as you suggested

Reply:
#18537